ESQL: Fix FieldAttribute name usage in InferNonNullAggConstraint

alex-spies opened the pull request:
Fix [#127521](https://github.com/elastic/elasticsearch/issues/127521)

I'm starting to fix a couple mis-usages of `FieldAttribute#name` (which is **not** always the actual field name!).

Because mis-usage of this method is currently common, I think we should squash this at compile time where possible. For this, I'm introducing a dedicated wrapper record `FieldName` to be used where field names are really required.

I'm starting with the biggest offender, `SearchStats`, and make this always require a `FieldName`, not a `String` - and make `FieldAttribute#fieldName` handily return an instance of `FieldName` so users of `SearchStats` don't accidentally use the return value of `FieldAttribute#name`.

Just changing the `SearchStats` API by requiring this wrapper automatically forced me to fix the bug in InferNonNullAggConstraint from #127521. It also surfaced another bug, c.f. https://github.com/elastic/elasticsearch/issues/128905.

alex-spies committed:
Make SearchStats use a dedicated FieldName class

Using String allowed bugs to slip through where we used the raw
attribute name instead of the actual field name.

alex-spies committed:
Fix InferNonNullAggConstraint

For union types, we should use the actual field name, not the attribute
name.

alex-spies committed:
Test fix for InferNonNullAggConstraint

alex-spies committed:
Remove TODO

Let's not introduce FieldName into the compute package - this one should
have less ambiguity whether a string might be a field name or not.

alex-spies committed:
Update TODO comment

elasticsearchmachine commented:
Hi @alex-spies, I've created a changelog YAML for you.

alex-spies committed:
Update docs/changelog/128910.yaml

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/local/InferNonNullAggConstraint.java:
```diff
@@ -55,7 +55,7 @@ protected LogicalPlan rule(Aggregate aggregate, LocalLogicalOptimizerContext con
                 // ignore literals (e.g. COUNT(1))
                 // make sure the field exists at the source and is indexed (not runtime)
-                if (field.foldable() == false && field instanceof FieldAttribute fa && stats.isIndexed(fa.name())) {
+                if (field.foldable() == false && field instanceof FieldAttribute fa && stats.isIndexed(fa.fieldName())) {
```
Yup, this is the actual fix.

alex-spies replied:
Caution!

`field.field().getName()` is wrong in case of subfields! The `EsField` doesn't know about the full path, so this is wrong without also using `field.parentName()`.

I hope this gets easier with the updated javadoc. :/

alex-spies committed:
Update test expectations

fang-xing-esql reviewed (approved):
Thank you @alex-spies, LGTM! There are a couple of things that I can think of:

The first is more of a nice-to-have. Since the rule was changed specifically to update the field name in the generated `IsNotNull` predicate, it might be helpful to add a test in `LocalLogicalPlanOptimizerTests` to verify this behavior, if we don’t already have one yet. As this should not affect the final results, and this is kind of a performance optimization.

The second is that there are quite a few places where `FieldName.toString()` is being called. I was wondering if there might be a better alternative. One reason is that it’s being called frequently. Another reason is that `toString()` is a general-purpose method. I was thinking to call it like `FieldName.name()`, however that feels a bit redundant. Maybe it doesn't worth overthinking, and it's best to leave it(`toString`) as-is unless there's a better naming suggestion from others.

alex-spies committed:
Do not override toString

Just use the attribute accessor method. Rename it to `string` to not
repeat fieldName.name() (we know it's a name) but rather express that
this is the string representation.

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into fix-fieldattribute-name-usage

alex-spies committed:
Fix conflicts with upstream changes to main

craigtaverner reviewed (approved):
Approved, but I did have some questions and concerns, particularly with creating new FieldName instances in the LucenePushdownPredicates class.

alex-spies commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/FieldAttribute.java:
```diff
@@ -27,14 +27,24 @@
+ *
+ * To adequately represent e.g. union types, the name of the attribute can be altered because we may have multiple synthetic field
+ * attributes that really belong to the same underlying field. For instance, if a multi-typed field is used both as {@code field::string}
+ * and {@code field::ip}, we'll generate 2 field attributes called {@code $$field$converted_to$string} and {@code $$field$converted_to$ip}
```
We do generate more than 1; I just confirmed this locally with the example given in the javadoc:
```
Eval[[TOIPLEADINGZEROSREJECTED(?field) AS field1#24, TOSTRING(?field) AS field2#27]] ! Project[[!field, field1{r}#24, field2{r}#27]]
\_UnresolvedRelation[test*]                                                          ! \_Limit[1000[INTEGER],false]
                                                                                     !   \_Eval[[$$field$converted_to$ip{f$}#29 AS field1#24, $$field$converted_to$keyword{f$}#30 AS field2#27]]
                                                                                     !     \_EsRelation[test*][!field, $$field$converted_to$ip{f$}#29, $$field$con..]
```

alex-spies commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/FieldAttribute.java:
```diff
@@ -27,14 +27,24 @@
+     * A field name, as found in the mapping. Includes the whole path from the root of the document.
+     * Implemented as a wrapper around {@link String} to distinguish from the attribute name (which sometimes differs!) at compile time.
+     */
+    public record FieldName(String string) {};
```
I had this initially, but then you have `attr.fieldName().name()` everywhere - which kinda loses meaning although it's correct.

Do we know of a precedent where we used a wrapper record before? In Rust, it'd be standard to call the wrapped attribute `inner` or `base`, but this is not Rust. ^^"

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/LucenePushdownPredicates.java:
```diff
@@ -126,32 +126,36 @@ public boolean canUseEqualityOnSyntheticSourceDelegate(FieldAttribute attr, Stri
             @Override
             public boolean hasExactSubfield(FieldAttribute attr) {
-                return stats.hasExactSubfield(attr.name());
+                return stats.hasExactSubfield(new FieldAttribute.FieldName(attr.name()));
```
This fails the union type tests - I didn't want to fix it in this PR but opened  https://github.com/elastic/elasticsearch/issues/128905

alex-spies replied:
Applies also to the instances below.

alex-spies committed:
Merge branch 'main' into fix-fieldattribute-name-usage

alex-spies committed:
Merge branch 'main' into fix-fieldattribute-name-usage

alex-spies committed:
Merge branch 'main' into fix-fieldattribute-name-usage

alex-spies commented:
CI is green except for 1 test failure in the Elasticsearch Serverless Checks, which is unrelated and since has been muted on Serverless. This is safe to merge.
